apps/prairielearn/src/lib/server-jobs-2.ts

```diff
@@ -0,0 +1,184 @@
+import AnsiUp from 'ansi_up';
+import execa = require('execa');
+import { z } from 'zod';
+import * as Sentry from '@prairielearn/sentry';
+import { logger } from '@prairielearn/logger';
+import { loadSqlEquiv, queryAsync, queryValidatedOneRow } from '@prairielearn/postgres';
+
+import { chalk } from './chalk';
+import serverJobs = require('./server-jobs');
+import socketServer = require('./socket-server');
+
+const sql = loadSqlEquiv(__filename);
+
+interface CreateServerJobOptions {
+  courseId?: string;
+  courseInstanceId?: string;
+  courseRequestId?: string;
+  assessmentId?: string;
+  userId?: string;
+  authnUserId?: string;
+  type: string;
+  description: string;
+}
+
+interface ServerJobExecOptions {
+  cwd?: string;
+  env?: NodeJS.ProcessEnv;
+}
+
+type ServerJobContext = Omit<ServerJob, 'execute' | 'executeInBackground'>;
+type JobExecutionFunction = (context: ServerJobContext) => Promise<void>;
+
+class ServerJob {
+  public jobSequenceId: string;
+  public jobId: string;
+  private started = false;
+  private finished = false;
+  public output = '';
+
+  constructor(jobSequenceId: string, jobId: string) {
+    this.jobSequenceId = jobSequenceId;
+    this.jobId = jobId;
+  }
+
+  error(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  warn(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  info(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  verbose(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  async exec(file: string, args: string[] = [], options: ServerJobExecOptions = {}): Promise<void> {
+    const msg = `$ ${file} ${args.join(' ')}\n`;
+    this.addToOutput(chalk.greenBright(msg));
+    const proc2 = execa(file, args, {
+      ...options,
+      reject: false,
+      all: true,
+    });
+
+    proc2.all?.setEncoding('utf-8');
+    proc2.all?.on('data', (data) => {
+      this.addToOutput(data);
+    });
+
+    const result = await proc2;
+    if (result.exitCode !== 0) {
+      throw new Error(`Process exited with code ${result.exitCode}`);
```

Comment:
This code doesn't log `result.signal` - should it?

Reply:
Seems like it wouldn't hurt?

Reply:
In 90f94d6 I changed it to use Execa's own error, which properly logs both exit codes and signals.